manage/config-dialog.js

```diff
@@ -286,7 +316,33 @@ function configDialog(style) {
   }
 
   function updateVarOnChange() {
-    this.va.value = this.type !== 'checkbox' ? this.value : this.checked ? '1' : '0';
+    if (this.type === 'number') {
+      this.value = this.va.value = clampValue(this.value, this.va);
+    } else if (this.type === 'range') {
+      this.va.value = Number(this.value);
+      $('.current-value', this.closest('.config-range')).textContent = this.va.value + (this.va.units || '');
+    } else {
+      this.va.value = this.type !== 'checkbox' ? this.value : this.checked ? '1' : '0';
+    }
+  }
+
+  // Clamp input[type=number] to a valid range
+  function clampValue(value, va, break = false) {
+    if (isNumber(va.min) && value < va.min) {
+      return va.min;
+    }
+    if (isNumber(va.max) && value > va.max) {
+      return va.max;
+    }
+    // Don't restrict to integer values if step is undefined.
+    if (!isNumber(va.step) || break) {
+      return value;
+    }
+    const step = va.step || 1;
+    const scale = 10 ** (step.toString().split('.')[1] || '').length;
+    value = Math.round((value * scale) / (step * scale)) * (step * scale) / scale;
+    // clamp modified value; skip step check
+    return clampValue(value, va, true);
```

Comment:
This is a bad idea. It makes the first min/max check useless. Let's move L341-L343 to L331 so the step validation is run before min/max validation.

Reply:
Yeah, I thought that breaking out of the recursion would be sufficient, but your idea is better.